### PR TITLE
Mid price

### PR DESCRIPTION
**Pull request by atvanguard**
## Why this should be merged
- Use mid price instead of last price for determining pnl. Corresponding contracts [PR](https://github.com/hubble-exchange/hubble-protocol/pull/205).

### CODE REVIEW

File: plugin/evm/orderbook/hubbleutils/margin_math.go [hunk 1]
```diff
@@ -90,7 +100,7 @@ func GetPositionMetadata(price *big.Int, openNotional *big.Int, size *big.Int, m
 }
 
 func GetNotionalPosition(price *big.Int, size *big.Int) *big.Int {
-	return big.NewInt(0).Abs(Div1e18(Mul(size, price)))
+	return big.NewInt(0).Abs(Div1e18(Mul(price, size)))
```
Comment (shubhamgoyal42) [inline plugin/evm/orderbook/hubbleutils/margin_math.go:103]: does the ordering make any difference, readability-wise?

File: plugin/evm/orderbook/liquidations.go [hunk 2]
```diff
@@ -19,17 +19,17 @@ type LiquidablePosition struct {
...
-	margin := new(big.Int).Sub(getNormalisedMargin(trader, assets), pendingFunding)
-	notionalPosition, unrealizePnL := getTotalNotionalPositionAndUnrealizedPnl(trader, margin, hu.Maintenance_Margin, oraclePrices, lastPrices, markets)
-	if notionalPosition.Sign() == 0 {
-		return big.NewInt(math.MaxInt64)
+func calcMarginFraction(trader *Trader, hState *hu.HubbleState) *big.Int {
+	userState := &hu.UserState{
+		Positions:      translatePositions(trader.Positions),
+		Margins:        getMargins(trader, len(hState.Assets)),
+		PendingFunding: getTotalFunding(trader, hState.ActiveMarkets),
+		ReservedMargin: trader.Margin.Reserved,
```
Comment (shubhamgoyal42) [inline plugin/evm/orderbook/liquidations.go:30]: value should be copied (`bigInt.Set`)

File: plugin/evm/orderbook/memory_database.go [hunk 3]
```diff
@@ -924,26 +916,30 @@ func (db *InMemoryDatabase) GetNaughtyTraders(oraclePrices map[Market]*big.Int,
...
 
 	for addr, trader := range db.TraderMap {
-		pendingFunding := getTotalFunding(trader, markets)
-		marginFraction := calcMarginFraction(trader, pendingFunding, assets, oraclePrices, db.LastPrice, markets)
-		if marginFraction.Cmp(db.configService.getMaintenanceMargin()) == -1 {
+		userState := &hu.UserState{
+			Positions:      translatePositions(trader.Positions),
+			Margins:        getMargins(trader, len(hState.Assets)),
+			PendingFunding: getTotalFunding(trader, hState.ActiveMarkets),
+			ReservedMargin: trader.Margin.Reserved,
```
Comment (shubhamgoyal42) [inline plugin/evm/orderbook/memory_database.go:923]: value should be copied

File: plugin/evm/orderbook/memory_database_test.go [hunk 4]
```diff
@@ -419,14 +419,22 @@ func TestGetCancellableOrders(t *testing.T) {
...
+	hState := &hu.HubbleState{
+		Assets:             assets,
+		OraclePrices:       priceMap,
+		MidPrices:          inMemoryDatabase.GetLastPrices(),
+		ActiveMarkets:      []Market{market},
+		MinAllowableMargin: inMemoryDatabase.configService.getMinAllowableMargin(),
+		MaintenanceMargin:  inMemoryDatabase.configService.getMaintenanceMargin(),
+	}
+	marginFraction := calcMarginFraction(_trader, hState)
+	// marginFraction := calcMarginFraction(_trader, big.NewInt(0), assets, priceMap, inMemoryDatabase.GetLastPrices(), []Market{market})
```
Comment (shubhamgoyal42) [inline plugin/evm/orderbook/memory_database_test.go:431]: can be removed

File: plugin/evm/orderbook/service.go [hunk 5]
```diff
@@ -126,8 +137,9 @@ func (api *OrderBookAPI) GetDebugData(ctx context.Context, trader string) GetDeb
 		response.ReservedMargin[addr] = trader.Margin.Reserved
 	}
 
-	response.LastPrice = lastPrices
+	response.LastPrice = api.db.GetLastPrices()
```
Comment (shubhamgoyal42) [inline plugin/evm/orderbook/service.go:140]: this function in memory db should return a copy of LastPrice

File: plugin/evm/orderbook/hubbleutils/margin_math.go (hunk 1 above)
Comment (atvanguard) [reply plugin/evm/orderbook/hubbleutils/margin_math.go:103]: hehe just writing in the same order of given arguments

**Review by shubhamgoyal42 — APPROVED** [review]

**Pull request merged**